### ISSUE TITLE
gen_keybindings macro doesn't work properly if run_internal is provided before run_external

### ISSUE THREAD

**Issue by yohannd1** [sminez/penrose#23]
**Describe the bug**
I had to miss a lot of details on the title, so let me explain.
I've noticed that, in the `gen_keybindings` macro, if any keybinding entry with a `run_internal` action is provided before a keybinding entry with a `run_external` action, the macro seems to generate code that results in closure type mismatches.

**Working Example**
First of all, a simple working `main.rs` example:
```
#[macro_use]
extern crate penrose;

use penrose::{Config, Forward, WindowManager, XcbConnection};

fn main() {
    let config = Config::default();
    let key_bindings = gen_keybindings! {
        "a" => run_external!("something"),
        "b" => run_internal!(cycle_client, Forward);

        forall_workspaces: config.workspaces => {
            "M-{}" => focus_workspace,
            "M-S-{}" => client_to_workspace,
        }
    };

    let conn = XcbConnection::new();
    let mut wm = WindowManager::init(config, &conn);
    wm.grab_keys_and_run(key_bindings);
}
```

**Non-working example**
```
#[macro_use]
extern crate penrose;

use penrose::{Config, Forward, WindowManager, XcbConnection};

fn main() {
    let config = Config::default();
    let key_bindings = gen_keybindings! {
        "b" => run_internal!(cycle_client, Forward);

        forall_workspaces: config.workspaces => {
            "M-{}" => focus_workspace,
            "M-S-{}" => client_to_workspace,
        }
    };

    let conn = XcbConnection::new();
    let mut wm = WindowManager::init(config, &conn);
    wm.grab_keys_and_run(key_bindings);
}
```
Compiler error message:
```
error[E0308]: mismatched types
  --> src/main.rs:8:24
   |
8  |       let key_bindings = gen_keybindings! {
   |  ________________________^
9  | |         "b" => run_internal!(cycle_client, Forward);
10 | |
11 | |         forall_workspaces: config.workspaces => {
...  |
14 | |         }
15 | |     };
   | |_____^ expected closure, found a different closure
   |
   = note: expected closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]`
              found closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10 i:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace
for more info)

error[E0308]: mismatched types
  --> src/main.rs:8:24
   |
8  |       let key_bindings = gen_keybindings! {
   |  ________________________^
9  | |         "b" => run_internal!(cycle_client, Forward);
10 | |
11 | |         forall_workspaces: config.workspaces => {
...  |
14 | |         }
15 | |     };
   | |_____^ expected closure, found a different closure
   |
   = note: expected closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]`
              found closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10 i:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace
for more info)

error[E0308]: mismatched types
  --> src/main.rs:19:26
   |
9  |         "b" => run_internal!(cycle_client, Forward);
   |                ------------------------------------ the found closure
...
19 |     wm.grab_keys_and_run(key_bindings);
   |                          ^^^^^^^^^^^^ expected trait object `dyn std::ops::FnMut`, found closure
   |
   = note: expected struct `std::collections::HashMap<_, std::boxed::Box<(dyn for<'r, 's> std::ops::FnMut(&'r mut penrose::WindowManager<'s>) + 'static)>>`
              found struct `std::collections::HashMap<_, std::boxed::Box<[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]>>`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `penrose-debugging`.

To learn more, run the command again with --verbose.
```
If simply swapping the `run_external` and `run_internal` entries' places, the error is similar, but slightly different:
```
error[E0308]: mismatched types
  --> src/main.rs:10:16
   |
9  |         "b" => run_internal!(cycle_client, Forward),
   |                ------------------------------------ the expected closure
10 |         "a" => run_external!("something");
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected closure, found trait object `dyn std::ops::FnMut`
   |
   = note: expected struct `std::boxed::Box<[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]>`
              found struct `std::boxed::Box<(dyn for<'r, 's> std::ops::FnMut(&'r mut penrose::WindowManager<'s>) + 'static)>`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace
for more info)

error[E0308]: mismatched types
  --> src/main.rs:8:24
   |
8  |       let key_bindings = gen_keybindings! {
   |  ________________________^
9  | |         "b" => run_internal!(cycle_client, Forward),
10 | |         "a" => run_external!("something");
11 | |
...  |
15 | |         }
16 | |     };
   | |_____^ expected closure, found a different closure
   |
   = note: expected closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]`
              found closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10 i:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace
for more info)

error[E0308]: mismatched types
  --> src/main.rs:8:24
   |
8  |       let key_bindings = gen_keybindings! {
   |  ________________________^
9  | |         "b" => run_internal!(cycle_client, Forward),
10 | |         "a" => run_external!("something");
11 | |
...  |
15 | |         }
16 | |     };
   | |_____^ expected closure, found a different closure
   |
   = note: expected closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]`
              found closure `[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10 i:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace
for more info)

error[E0308]: mismatched types
  --> src/main.rs:20:26
   |
9  |         "b" => run_internal!(cycle_client, Forward),
   |                ------------------------------------ the found closure
...
20 |     wm.grab_keys_and_run(key_bindings);
   |                          ^^^^^^^^^^^^ expected trait object `dyn std::ops::FnMut`, found closure
   |
   = note: expected struct `std::collections::HashMap<_, std::boxed::Box<(dyn for<'r, 's> std::ops::FnMut(&'r mut penrose::WindowManager<'s>) + 'static)>>`
              found struct `std::collections::HashMap<_, std::boxed::Box<[closure@/home/yohanan/.cache/cargo/git/checkouts/penrose-073b20e3311c97ea/e570756/src/core/macros.rs:27:18: 29:10]>>`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `penrose-debugging`.

To learn more, run the command again with --verbose.
```

**Desktop (please complete the following information):**
 - OS: GNU/Linux
 - Distribution: Arch
 - Version: 5.7.8-arch1-1

**Additional context**
Since it's a compile-time error, there's no additional context I could place here.

**Comment by sminez** [comment]
Ah, thanks for the detail! I think I know what this will be: I got similar issues previously but I thought I'd resolved them... obviously not. I should ha e some time this evening to take a look

**Comment by sminez** [comment]
Fixed in #24
